File: internal/wire/codec.go
package wire

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
)

// ErrEmptyFrame is returned by Decode when given a zero-length frame.
var ErrEmptyFrame = errors.New("wire: empty frame")

// rawOpcodes are the opcodes whose payload is written verbatim rather than
// JSON-marshalled.
func isRawOpcode(op Opcode) bool {
	return op == OpInput || op == OpOutput
}

// Encode writes op as byte 0 followed by payload. For OpInput and OpOutput,
// payload must be a []byte and is written verbatim. Every other opcode's
// payload is JSON-marshalled; pass nil for the empty-payload opcodes
// (OpPing, OpPong).
func Encode(op Opcode, payload any) ([]byte, error) {
	if isRawOpcode(op) {
		raw, ok := payload.([]byte)
		if !ok {
			return nil, fmt.Errorf("wire: encode %s: payload must be []byte, got %T", op, payload)
		}
		frame := make([]byte, 1+len(raw))
		frame[0] = byte(op)
		copy(frame[1:], raw)
		return frame, nil
	}

	if payload == nil {
		return []byte{byte(op)}, nil
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("wire: encode %s: %w", op, err)
	}
	frame := make([]byte, 1+len(body))
	frame[0] = byte(op)
	copy(frame[1:], body)
	return frame, nil
}

// Decode splits frame into its opcode and remaining payload bytes. It
// performs no JSON parsing beyond validating, for known JSON opcodes, that
// the payload (when non-empty) is syntactically valid JSON. Decode never
// panics.
func Decode(frame []byte) (Opcode, []byte, error) {
	if len(frame) == 0 {
		return 0, nil, ErrEmptyFrame
	}

	op := Opcode(frame[0])
	payload := frame[1:]

	if !knownOpcode(op) {
		return 0, nil, fmt.Errorf("wire: unknown opcode 0x%02x", frame[0])
	}

	if isRawOpcode(op) {
		return op, payload, nil
	}

	if len(payload) > 0 && !json.Valid(payload) {
		return 0, nil, fmt.Errorf("wire: decode %s: invalid JSON payload", op)
	}

	return op, payload, nil
}

// knownOpcode reports whether b falls within the valid opcode byte range
// (0x00-0x06) shared by both the Client->Server and Server->Client tables.
// It cannot tell the two tables apart on the byte value alone — see the
// Opcode.String doc comment — so callers that need direction-specific
// validation must check against the specific opcode set they expect.
func knownOpcode(op Opcode) bool {
	return byte(op) <= byte(OpPong)
}

// EncodeTo encodes op/payload and writes the resulting frame to w.
func EncodeTo(w io.Writer, op Opcode, payload any) error {
	frame, err := Encode(op, payload)
	if err != nil {
		return err
	}
	_, err = w.Write(frame)
	return err
}

// DecodeFrom reads all of r and decodes it as a single frame. It is
// intended for tests and tooling that hand it exactly one message's worth
// of bytes; it is not used on the live WebSocket path, where the transport
// already delivers one frame per message.
func DecodeFrom(r io.Reader) (Opcode, []byte, error) {
	frame, err := io.ReadAll(r)
	if err != nil {
		return 0, nil, fmt.Errorf("wire: decode from reader: %w", err)
	}
	return Decode(frame)
}

// DecodeResize decodes a Resize payload and clamps it into range.
func DecodeResize(payload []byte) (Resize, error) {
	var r Resize
	if err := json.Unmarshal(payload, &r); err != nil {
		return Resize{}, fmt.Errorf("wire: decode Resize: %w", err)
	}
	return r.Clamp(), nil
}

// DecodeHello decodes a Hello payload.
func DecodeHello(payload []byte) (Hello, error) {
	var h Hello
	if err := json.Unmarshal(payload, &h); err != nil {
		return Hello{}, fmt.Errorf("wire: decode Hello: %w", err)
	}
	return h, nil
}

// DecodeRoster decodes a Roster payload.
func DecodeRoster(payload []byte) (Roster, error) {
	var r Roster
	if err := json.Unmarshal(payload, &r); err != nil {
		return Roster{}, fmt.Errorf("wire: decode Roster: %w", err)
	}
	return r, nil
}

// DecodeState decodes a State payload and validates State.State against
// the live/detached/closed enum.
func DecodeState(payload []byte) (State, error) {
	var s State
	if err := json.Unmarshal(payload, &s); err != nil {
		return State{}, fmt.Errorf("wire: decode State: %w", err)
	}
	switch s.State {
	case SessionStateLive, SessionStateDetached, SessionStateClosed:
	default:
		return State{}, fmt.Errorf("wire: decode State: invalid state %q", s.State)
	}
	return s, nil
}

// DecodeExit decodes an Exit payload.
func DecodeExit(payload []byte) (Exit, error) {
	var e Exit
	if err := json.Unmarshal(payload, &e); err != nil {
		return Exit{}, fmt.Errorf("wire: decode Exit: %w", err)
	}
	return e, nil
}

// DecodeNotice decodes a Notice payload and validates Notice.Level against
// the info/warn/error enum.
func DecodeNotice(payload []byte) (Notice, error) {
	var n Notice
	if err := json.Unmarshal(payload, &n); err != nil {
		return Notice{}, fmt.Errorf("wire: decode Notice: %w", err)
	}
	switch n.Level {
	case NoticeLevelInfo, NoticeLevelWarn, NoticeLevelError:
	default:
		return Notice{}, fmt.Errorf("wire: decode Notice: invalid level %q", n.Level)
	}
	return n, nil
}

File: web/src/protocol/codec.ts
import { ClientOpcode, ServerOpcode } from "./opcodes";
import type { Exit, Hello, Notice, Resize, Roster, State } from "./types";

/** Encodes a Resize frame: opcode byte + JSON payload. */
export function encodeResize(resize: Resize): ArrayBuffer {
  return encodeJSON(ClientOpcode.Resize, resize);
}

/** Encodes an Input frame: opcode byte + raw bytes verbatim. */
export function encodeInput(data: Uint8Array): ArrayBuffer {
  return encodeRaw(ClientOpcode.Input, data);
}

/** Encodes a Ping frame: opcode byte, empty payload. */
export function encodePing(): ArrayBuffer {
  return encodeRaw(ClientOpcode.Ping, new Uint8Array(0));
}

function encodeRaw(op: number, data: Uint8Array): ArrayBuffer {
  const frame = new Uint8Array(1 + data.length);
  frame[0] = op;
  frame.set(data, 1);
  return frame.buffer;
}

function encodeJSON(op: number, payload: unknown): ArrayBuffer {
  const body = new TextEncoder().encode(JSON.stringify(payload));
  return encodeRaw(op, body);
}

export type DecodedFrame =
  | { op: ServerOpcode.Output; data: Uint8Array }
  | { op: ServerOpcode.Hello; data: Hello }
  | { op: ServerOpcode.Roster; data: Roster }
  | { op: ServerOpcode.State; data: State }
  | { op: ServerOpcode.Exit; data: Exit }
  | { op: ServerOpcode.Notice; data: Notice }
  | { op: ServerOpcode.Pong }
  | { op: "unknown"; opcode: number };

/**
 * Decodes one Server -> Client frame. Never throws: a truncated or
 * malformed frame is reported as an error result rather than propagated,
 * and an opcode outside the known set decodes to "unknown" so the caller
 * can ignore it instead of treating it as fatal (SPEC.md §5.2).
 */
export function decodeFrame(buf: ArrayBuffer): DecodedFrame | { op: "error"; message: string } {
  if (buf.byteLength === 0) {
    return { op: "error", message: "ditty: empty frame" };
  }
  const bytes = new Uint8Array(buf);
  const opcode = bytes[0];
  const payload = bytes.subarray(1);

  switch (opcode) {
    case ServerOpcode.Output:
      return { op: ServerOpcode.Output, data: payload };
    case ServerOpcode.Pong:
      return { op: ServerOpcode.Pong };
    case ServerOpcode.Hello:
      return decodeJSON<ServerOpcode.Hello, Hello>(ServerOpcode.Hello, payload);
    case ServerOpcode.Roster:
      return decodeJSON<ServerOpcode.Roster, Roster>(ServerOpcode.Roster, payload);
    case ServerOpcode.State:
      return decodeJSON<ServerOpcode.State, State>(ServerOpcode.State, payload);
    case ServerOpcode.Exit:
      return decodeJSON<ServerOpcode.Exit, Exit>(ServerOpcode.Exit, payload);
    case ServerOpcode.Notice:
      return decodeJSON<ServerOpcode.Notice, Notice>(ServerOpcode.Notice, payload);
    default:
      return { op: "unknown", opcode };
  }
}

function decodeJSON<Op extends ServerOpcode, T>(
  op: Op,
  payload: Uint8Array,
): { op: Op; data: T } | { op: "error"; message: string } {
  try {
    const text = new TextDecoder().decode(payload);
    return { op, data: JSON.parse(text) as T };
  } catch (err) {
    return { op: "error", message: `ditty: malformed frame for opcode ${op}: ${String(err)}` };
  }
}
